appcues/src/main/java/com/appcues/ui/composables/AppcuesComposition.kt

```diff
@@ -91,44 +95,63 @@ private fun BoxScope.ComposeLastRenderingState(state: Rendering) {
     }
 
     with(state.stepContainer) {
-        // apply backdrop traits
-        ApplyBackgroundDecoratingTraits(list = backdropDecoratingTraits.toMutableList())
-
-        // create wrapper
-        contentWrappingTrait.WrapContent { hasFixedHeight, contentPadding ->
-            Box(contentAlignment = Alignment.TopCenter) {
-                ApplyUnderlayContainerTraits(this)
-
-                // Apply content holder trait
-                with(contentHolderTrait) {
-                    // create object that will passed down to CreateContentHolder
-                    ContainerPages(
-                        pageCount = steps.size,
-                        currentPage = state.position,
-                        composePage = { index ->
-                            with(steps[index]) {
-                                CompositionLocalProvider(LocalAppcuesActions provides actions) {
-                                    // used to get the padding values from step decorating trait and apply to the Column
-                                    val density = LocalDensity.current
-                                    val stepDecoratingPadding = remember(this) { StepDecoratingPadding(density) }
-
-                                    ApplyUnderlayStepTraits(this@Box, stepDecoratingPadding)
-
-                                    ComposeStepContent(index, hasFixedHeight, contentPadding, stepDecoratingPadding)
-
-                                    ApplyOverlayStepTraits(this@Box, stepDecoratingPadding)
+        val backdropDecoratingTraits = remember(state.position) { mutableStateOf(steps[state.position].backdropDecoratingTraits) }
+        val containerDecoratingTraits = remember(state.position) { mutableStateOf(steps[state.position].containerDecoratingTraits) }
+        val metadataSettingTraits = remember(state.position) { mutableStateOf(steps[state.position].metadataSettingTraits) }
+        val previousStepMetaData = remember { mutableStateOf(AppcuesStepMetadata()) }
+        val stepMetadata = remember(metadataSettingTraits.value) {
+
+            val actual = hashMapOf<String, Any?>().apply {
+                metadataSettingTraits.value.forEach {
+                    putAll(it.produceMetadata())
+                }
+            }
+
+            mutableStateOf(AppcuesStepMetadata(previous = previousStepMetaData.value.actual, actual = actual)).also {
+                previousStepMetaData.value = it.value
+            }
+        }
+
+        CompositionLocalProvider(LocalAppcuesStepMetadata provides stepMetadata.value) {
```

Comment:
All traits, steps can access the most recent stepMetadata value, that contains the actual and possibly a previous value of that same key